File: server/game/protocol.go
package game

import "encoding/json"

type Suit string

const (
	Clubs    Suit = "clubs"
	Diamonds Suit = "diamonds"
	Hearts   Suit = "hearts"
	Spades   Suit = "spades"
	Joker    Suit = "joker"
)

type Card struct {
	ID   string `json:"id"`
	Rank int    `json:"rank"`
	Suit Suit   `json:"suit"`
}

type CardRef struct {
	PlayerID string `json:"playerId"`
	Slot     int    `json:"slot"`
}

type ClientMessage struct {
	Type          string  `json:"type"`
	Slot          int     `json:"slot"`
	Target        CardRef `json:"target"`
	First         CardRef `json:"first"`
	Second        CardRef `json:"second"`
	EventID       int     `json:"eventId"`
	SourceSlot    int     `json:"sourceSlot"`
	JoinNextRound bool    `json:"joinNextRound"`
}

type CardSlotView struct {
	Slot     int   `json:"slot"`
	Occupied bool  `json:"occupied"`
	Card     *Card `json:"card,omitempty"`
}

type PlayerView struct {
	ID           string         `json:"id"`
	Name         string         `json:"name"`
	Connected    bool           `json:"connected"`
	Cards        []CardSlotView `json:"cards"`
	InitialReady *bool          `json:"initialReady,omitempty"`
	Score        *int           `json:"score,omitempty"`
}

type RosterPlayerView struct {
	ID               string `json:"id"`
	Name             string `json:"name"`
	Connected        bool   `json:"connected"`
	JoiningNextRound bool   `json:"joiningNextRound"`
}

type RevealCard struct {
	Target CardRef `json:"target"`
	Card   Card    `json:"card"`
}

type RevealView struct {
	Kind  string       `json:"kind"`
	Cards []RevealCard `json:"cards"`
}

type GiftView struct {
	SlapperID string  `json:"slapperId"`
	Target    CardRef `json:"target"`
}

type PublicPeekView struct {
	ViewerID string  `json:"viewerId"`
	Target   CardRef `json:"target"`
}

type ActionView struct {
	ID      int      `json:"id"`
	Kind    string   `json:"kind"`
	ActorID string   `json:"actorId"`
	First   *CardRef `json:"first,omitempty"`
	Second  *CardRef `json:"second,omitempty"`
	Target  *CardRef `json:"target,omitempty"`
}

type Snapshot struct {
	Type             string             `json:"type"`
	RoomID           string             `json:"roomId"`
	You              Identity           `json:"you"`
	YouRole          string             `json:"youRole"`
	NextRoundJoined  bool               `json:"nextRoundJoined"`
	NextRoundFull    bool               `json:"nextRoundFull"`
	NextRoundPlayers []RosterPlayerView `json:"nextRoundPlayers"`
	WaitingPlayers   []RosterPlayerView `json:"waitingPlayers"`
	Phase            Phase              `json:"phase"`
	CurrentPlayerID  string             `json:"currentPlayerId,omitempty"`
	Players          []PlayerView       `json:"players"`
	DrawPileCount    int                `json:"drawPileCount"`
	HasDrawnCard     bool               `json:"hasDrawnCard"`
	DiscardTop       *Card              `json:"discardTop,omitempty"`
	DiscardEventID   int                `json:"discardEventId,omitempty"`
	DrawnCard        *Card              `json:"drawnCard,omitempty"`
	Reveal           *RevealView        `json:"reveal,omitempty"`
	PublicPeek       *PublicPeekView    `json:"publicPeek,omitempty"`
	Action           *ActionView        `json:"action,omitempty"`
	PendingGift      *GiftView          `json:"pendingGift,omitempty"`
	EndReason        string             `json:"endReason,omitempty"`
	WinnerIDs        []string           `json:"winnerIds,omitempty"`
}

type Identity struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type ErrorMessage struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

func EncodeError(code, message string) []byte {
	b, _ := json.Marshal(ErrorMessage{Type: "error", Code: code, Message: message})
	return b
}

File: shared/protocol.ts
export type Suit = "clubs" | "diamonds" | "hearts" | "spades" | "joker";

export interface Card {
  id: string;
  rank: number;
  suit: Suit;
}

export interface CardRef {
  playerId: string;
  slot: number;
}

export type GamePhase =
  | "lobby"
  | "initial_peek"
  | "await_draw"
  | "await_choice"
  | "await_self_peek"
  | "await_opponent_peek"
  | "reveal_self"
  | "reveal_opponent"
  | "await_swap"
  | "await_king_peek"
  | "reveal_king"
  | "await_gift"
  | "ended";

export interface CardSlotView {
  slot: number;
  occupied: boolean;
  card?: Card;
}

export interface PlayerView {
  id: string;
  name: string;
  connected: boolean;
  cards: CardSlotView[];
  initialReady?: boolean;
  score?: number;
}

export interface RosterPlayerView {
  id: string;
  name: string;
  connected: boolean;
  joiningNextRound: boolean;
  ready: boolean;
}

export interface RevealView {
  kind: "initial" | "self" | "opponent" | "king";
  cards: Array<{ target: CardRef; card: Card }>;
}

export interface GiftView {
  slapperId: string;
  target: CardRef;
}

export interface PublicPeekView {
  viewerId: string;
  target: CardRef;
}

export interface ActionView {
  id: number;
  kind: "swap" | "replace" | "discard" | "slap" | "gift";
  actorId: string;
  first?: CardRef;
  second?: CardRef;
  target?: CardRef;
}

export interface SnapshotMessage {
  type: "snapshot";
  roomId: string;
  you: { id: string; name: string };
  youRole: "active" | "spectator";
  nextRoundJoined: boolean;
  nextRoundFull: boolean;
  youReady: boolean;
  allReady: boolean;
  nextStarterId?: string;
  nextRoundPlayers: RosterPlayerView[];
  waitingPlayers: RosterPlayerView[];
  phase: GamePhase;
  currentPlayerId?: string;
  players: PlayerView[];
  drawPileCount: number;
  hasDrawnCard: boolean;
  discardTop?: Card;
  discardEventId?: number;
  drawnCard?: Card;
  reveal?: RevealView;
  publicPeek?: PublicPeekView;
  action?: ActionView;
  pendingGift?: GiftView;
  endReason?: string;
  winnerIds?: string[];
}

export interface ErrorMessage {
  type: "error";
  code: string;
  message: string;
}

export interface NoticeMessage {
  type: "notice";
  message: string;
}

export type ServerMessage = SnapshotMessage | ErrorMessage | NoticeMessage;

export type ClientMessage =
  | { type: "start_game" }
  | { type: "set_next_round"; joinNextRound: boolean }
  | { type: "set_ready"; ready: boolean }
  | { type: "acknowledge_initial" }
  | { type: "draw" }
  | { type: "replace"; slot: number }
  | { type: "discard_drawn" }
  | { type: "peek"; target: CardRef }
  | { type: "acknowledge_reveal" }
  | { type: "swap"; first: CardRef; second: CardRef }
  | { type: "slap"; eventId: number; target: CardRef }
  | { type: "gift"; sourceSlot: number }
  | { type: "call_end" };

export function cardScore(card: Pick<Card, "rank" | "suit">): number {
  if (card.rank === 0) return 0;
  if (card.rank === 13 && (card.suit === "hearts" || card.suit === "diamonds")) return -1;
  return card.rank;
}
